awslabs/aws-lambda-powertools-typescript
# feat(logger): add clear state functionality

## Description of your changes

The Logger utility is commonly initialized in the global scope, outside the handler function.
When you attach persistent log attributes through the `persistentLogAttributes` constructor option or via the `appendKeys`, `addPersistentLogAttributes` methods, this data is attached to the Logger instance.  

Due to the [Lambda Execution Context reuse](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-context.html), this means those persistent log attributes may be reused across invocations.

This PR introduces a new flag that developers can enable to "clear the state" across invocations.
If developers want to make sure that persistent attributes added **inside the handler function** code are not persisted across invocations, they can now set the parameter `clearState` as `true`  in the `injectLambdaContext` middleware or decorator.









### How to verify this change



See [Contributing guide](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/CONTRIBUTING.md) for the docs build. 




![carbon (3)](https://user-images.githubusercontent.com/47529391/170583330-d8938326-237c-47cd-8f5c-0f45ca5d52a5.png)


### Related issues, RFCs


[#482](https://github.com/awslabs/aws-lambda-powertools-typescript/issues/482)  

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist


### Breaking change checklist


---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
